### ISSUE TITLE
The ASB advancing incorrectly when waiting too long for redeem tx

### ISSUE THREAD

**Issue by da-kami** [comit-network/xmr-btc-swap#514]
## Scenario and Analysis

Both CLI and ASB redeemed successfully, but the ASB's Bitcoin redeem transaction needs significant time to get included because the network overheats (in the current case transaction fees spiked because Bitcoin is crashing).

This eventually results in a timeout on the ASB side:

```
2021-05-19 17:46:30 DEBUG Transaction txid=0b8fdcff62408252762f60c498f68a4ab8ee19dcc0bb6ade3eb95e9bbd4a6c5a status=unseen
2021-05-19 17:46:37 DEBUG Transaction txid=0b8fdcff62408252762f60c498f68a4ab8ee19dcc0bb6ade3eb95e9bbd4a6c5a status=in mempool
2021-05-19 17:47:10  WARN Lost connection. Error Connection error: I/O error: connection is closed peer=12D3KooWL61zVenQs5bgrzXE9tFiqKzRxgdKybViagBk8jUcxj7Y address=/ip4/127.0.0.1/tcp/50617
2021-05-19 18:03:26 DEBUG Got notification for new block block_height=684169
2021-05-19 18:03:26 DEBUG Transaction txid=213da99a0d090ae0065146592ff32d886b5262927c8e754dd0549f3146fb7527 status=confirmed with 4 blocks
2021-05-19 18:08:40 DEBUG Got notification for new block block_height=684170
2021-05-19 18:08:40 DEBUG Transaction txid=213da99a0d090ae0065146592ff32d886b5262927c8e754dd0549f3146fb7527 status=confirmed with 5 blocks
2021-05-20 06:13:36  WARN Failed to get status of script. Error Failed to subscribe to header notifications: Made one or multiple attempts, all errored:
	- Network is down (os error 50)
 txid=0b8fdcff62408252762f60c498f68a4ab8ee19dcc0bb6ade3eb95e9bbd4a6c5a
2021-05-20 06:13:36 ERROR swap{id=dcb9784e-6204-4eb3-98f4-8997768b3889}: error=Waiting for Bitcoin transaction finality failed with Failed while waiting for next status update! The redeem transaction was published, but it is not ensured that the transaction was included! You're screwed.
2021-05-20 06:13:36 ERROR Swap failed. Error Waiting for Bitcoin transaction finality failed with Failed while waiting for next status update! The redeem transaction was published, but it is not ensured that the transaction was included! You're screwed. swap_id=dcb9784e-6204-4eb3-98f4-8997768b3889

```

Since this is treated as unrecoverable error the ASB will end swap execution until restarted.

Upon restart the ASB tries to publish the transaction again (because it was unclear if the transaction was actually included). That obviously fails. Since the cancel timelock had expired the ASB then advances to trying to cancelling. When publishing the cancel transaction fails it assumes that the CLI has published it and waits for the refund transaction:

```
2021-05-20 07:58:28  INFO swap{id=dcb9784e-6204-4eb3-98f4-8997768b3889}: Advancing state state=encrypted signature is learned rate=NaN
2021-05-20 07:58:28 DEBUG swap{id=dcb9784e-6204-4eb3-98f4-8997768b3889}: Redeem tx fee spending_fee=7092
2021-05-20 07:58:28 ERROR swap{id=dcb9784e-6204-4eb3-98f4-8997768b3889}: Publishing the redeem transaction failed. Error Failed to broadcast Bitcoin redeem transaction 0b8fdcff62408252762f60c498f68a4ab8ee19dcc0bb6ade3eb95e9bbd4a6c5a: Electrum(Protocol(String("sendrawtransaction RPC error: {\"code\":-27,\"message\":\"Transaction already in block chain\"}")))
2021-05-20 07:58:29 DEBUG Connected to Kraken websocket API
2021-05-20 07:58:29 DEBUG Subscribed to updates for ticker
2021-05-20 07:58:33 DEBUG Transaction txid=213da99a0d090ae0065146592ff32d886b5262927c8e754dd0549f3146fb7527 status=unseen
2021-05-20 07:58:33 DEBUG Transaction txid=0b8fdcff62408252762f60c498f68a4ab8ee19dcc0bb6ade3eb95e9bbd4a6c5a status=unseen
2021-05-20 07:59:26 DEBUG Got notification for new block block_height=684244
2021-05-20 07:59:26 DEBUG Transaction txid=213da99a0d090ae0065146592ff32d886b5262927c8e754dd0549f3146fb7527 status=confirmed with 79 blocks
2021-05-20 07:59:26 DEBUG Transaction txid=0b8fdcff62408252762f60c498f68a4ab8ee19dcc0bb6ade3eb95e9bbd4a6c5a status=confirmed with 63 blocks
2021-05-20 07:59:26  INFO swap{id=dcb9784e-6204-4eb3-98f4-8997768b3889}: Advancing state state=cancel timelock is expired rate=0.00582500 BTC
2021-05-20 07:59:27 DEBUG swap{id=dcb9784e-6204-4eb3-98f4-8997768b3889}: Assuming cancel transaction is already broadcasted because: Failed to broadcast Bitcoin cancel transaction eae2c467792ee33e04c5cf1b14a13c37774c7d60a7c25cfdc835792af8c6ceb7: Electrum(Protocol(String("sendrawtransaction RPC error: {\"code\":-25,\"message\":\"bad-txns-inputs-missingorspent\"}")))
2021-05-20 07:59:27  INFO swap{id=dcb9784e-6204-4eb3-98f4-8997768b3889}: Advancing state state=btc is cancelled rate=0.00582500 BTC
2021-05-20 07:59:32 DEBUG Transaction txid=eae2c467792ee33e04c5cf1b14a13c37774c7d60a7c25cfdc835792af8c6ceb7 status=unseen
2021-05-20 07:59:32 DEBUG Transaction txid=2b3c9c75ec940445fd50eb28bde33132160818c022f10fd23dbbf883e37ddc13 status=unseen
```

Since refund will never happen the swap would end up in a "wrong" punish scenario:

```
2021-05-20 10:23:07  INFO swap{id=dcb9784e-6204-4eb3-98f4-8997768b3889}: Advancing state state=btc is punishable rate=0.00560800 BTC
2021-05-20 10:23:15  WARN swap{id=dcb9784e-6204-4eb3-98f4-8997768b3889}: Falling back to refund because punish transaction failed. Error Failed to broadcast Bitcoin punish transaction b110a6e15b71d3a488b0f8bc32699f0449e1fe515ab0723e7910b1f2ab0fe705: Electrum(AllAttemptsErrored([SharedIOError(Os { code: 54, kind: ConnectionReset, message: "Connection reset by peer" })]))
2021-05-20 10:23:18 ERROR swap{id=dcb9784e-6204-4eb3-98f4-8997768b3889}: error=Electrum(Protocol(String("missing transaction")))
2021-05-20 10:23:18 DEBUG A swap stopped without sending a transfer proof. Error request channel closed
2021-05-20 10:23:18 ERROR Swap failed. Error Electrum(Protocol(String("missing transaction"))) swap_id=dcb9784e-6204-4eb3-98f4-8997768b3889
2021-05-20 10:23:23 DEBUG Transaction txid=b110a6e15b71d3a488b0f8bc32699f0449e1fe515ab0723e7910b1f2ab0fe705 status=unseen
```

No money is lost in such a scenario, but it is a bit of an annoying scenario, because the swap can currently no be manually advanced to the correct state. The manual `redeem` command also tries to publish the redeem transaction which will result in an error. 
This effectively means that this swap will eventually end in a failing punish state, which will fail upon every restart (even though the swap was actually redeemed).  

## Possible Fixes

1. Transition to a model where we check if a transaction was included before publishing it. I think we should do that for all (Bitcoin) transactions. This helps with manual commands.
2. Introduce another state in the ASB state machine, after the redeem transaction was seen in mempool the ASB should never transition to the `cancel` path, but wait (indefinitely) for the redeem transaction to be finalized. We *can* wrap waiting for confirmations in large timeout that prints an error message so the ASB provider can react.

**Comment by thomaseizinger** [comment]
Agreed, there is a specific error code we should be able to listen for in case a transaction is already published.

I think in that case, the function should just be a idempotent and return successfully.

**Comment by bonomat** [comment]
Good analysis 👍

A solution could indeed be to either check if the transaction was already included and/or check for the error response on re-publishing. Note on that, bitcoind has a nice specific [error code](https://github.com/bitcoin/bitcoin/blob/58c07426326dd047c5fec8b4c1037b9c0da7a191/src/rpc/protocol.h#L48) but we use bdk through electrum to publish transactions and this error [code does](https://github.com/spesmilo/electrum/blob/de716ab07cf81c96658713026b03a3d3f1940c19/electrum/network.py#L1066-L1069) does not seem to be that useful because it could either mean: "everything is fine" or "hold on, your input tx does not exist!".  

Anyway, I'm wondering if we need to wait for transaction finality here (as defined in the config). 
As the log messages says: 
> [...] The redeem transaction was published, but it is not ensured that the transaction was included! You're screwed`

For the ASB there is not much we can do. We cannot republish with RBF because we need the other party's signature. In that case, we should just be "happy" if the transaction was seen in the mempool and proceed to a state _redeemed_. 
We should keep watching for the transaction to be confirmed and republish if it ever falls out of the mempool. Once we have our confirmations we can move on to a state _redeemed and included_ or so.

**Comment by rishflab** [comment]
> I propose to transition to a model where we check if a transaction was included before publishing it. I think we should do that for all (Bitcoin) transactions.

I think this is a good direction. Having "idempotent" transactions (for Monero as well) makes our state machine machine simpler and our code generally easier to reason about.

**Comment by thomaseizinger** [comment]
If I am not mistaken, electrum sends back the original RPC error as part of [this](https://docs.rs/electrum-client/0.7.0/electrum_client/enum.Error.html#variant.Protocol) error variant.

We could deserialize this into [`JsonRpcError`](https://docs.rs/jsonrpc_client/0.6.0/jsonrpc_client/struct.JsonRpcError.html) and match against the error code.

**Comment by bonomat** [comment]
Good idea. 

The question remains: what do we do with this information. 

You see in the logs above that there were two different error messages: 

```
Electrum(Protocol(String("sendrawtransaction RPC error: {\"code\":-27,\"message\":\"Transaction already in block chain\"}")))
```
Because the *same* transaction was already included. So this is fine if we republish the same transaction. 

```
Electrum(Protocol(String("sendrawtransaction RPC error: {\"code\":-25,\"message\":\"bad-txns-inputs-missingorspent\"}")
```

My interpretation is that this it is not clear what went wrong. E.g. it could be that we tried to publish `TxCancel` while `TxRedeem` was already in the mempool (as happened above) or it could be that the former transaction fell out somehow, e.g. `TxLock` is _gone_ due to a fork/orphaned block/etc. This would be really bad. 


The former error should be easy to handle, e.g. by just ignoring it. The latter one is a bit more tricky. We would need to figure out why the transaction failed, e.g. was `TxRedeem` included?

**Comment by da-kami** [comment]
> Good analysis 👍
> 
> A solution could indeed be to either check if the transaction was already included and/or check for the error response on re-publishing. Note on that, bitcoind has a nice specific [error code](https://github.com/bitcoin/bitcoin/blob/58c07426326dd047c5fec8b4c1037b9c0da7a191/src/rpc/protocol.h#L48) but we use bdk through electrum to publish transactions and this error [code does](https://github.com/spesmilo/electrum/blob/de716ab07cf81c96658713026b03a3d3f1940c19/electrum/network.py#L1066-L1069) does not seem to be that useful because at it seems like the same error could mean anything from: "everything is fine" to "hold on, your input tx does not exist!".

I think the important part is that broadcast should not re-broadcast - the [error](https://github.com/spesmilo/electrum/blob/de716ab07cf81c96658713026b03a3d3f1940c19/electrum/network.py#L1041) for that scenario is specific:

```
2021-05-20 07:58:28 ERROR swap{id=dcb9784e-6204-4eb3-98f4-8997768b3889}: Publishing the redeem transaction failed. Error Failed to broadcast Bitcoin redeem transaction 0b8fdcff62408252762f60c498f68a4ab8ee19dcc0bb6ade3eb95e9bbd4a6c5a: Electrum(Protocol(String("sendrawtransaction RPC error: {\"code\":-27,\"message\":\"Transaction already in block chain\"}")))
```

Scenarios where the transaction cannot be published because the output was already spent will have to be treated differently.

> 
> Anyway, I'm wondering if we need to wait for transaction finality here (as defined in the config).
> As the log messages says:
> 
> > [...] The redeem transaction was published, but it is not ensured that the transaction was included! You're screwed`
> 
> For the ASB there is not much we can do. We cannot republish with RBF because we need the other party's signature. In that case, we should just be "happy" if the transaction was seen in the mempool and proceed to a state _redeemed_.
> We should keep watching for the transaction to be confirmed and republish if it ever falls out of the mempool. Once we have our confirmations we can move on to a state _redeemed and included_ or so.

Introducing two separate states just shifts the problem to a different state. We can make the state more fine granular here, but I don't think it would help. 
If we don't wait for confirmations on redeem at all the problem is solved.
If we remove the timeout on waiting for confirmations the problem is solved as well. 
If we wait for a confirmations, and that waiting is bound by a timeout then the problem will exist no matter in what state - we would then have to consider re-publishing the transaction.

This error was meant for cases where there are network problems and the ASB provider should be notified that somehow we are not getting confirmations. 

On top of my proposal to check in broadcast that ensures the tx was not broadcasted yet, we can work on internal recovery during runtime eventually. This was described in this ticket https://github.com/comit-network/xmr-btc-swap/issues/375, but thought the manual commands would be enough for now.
I think they are - and with adding the check to broadcast they will be. 

We could additionally allow a longer timeout for redeem transaction, and we could also reduce the redeem confirmations.

**Comment by thomaseizinger** [comment]
`Electrum(Protocol(String("sendrawtransaction RPC error: {\"code\":-27,\"message\":\"Transaction already in block chain\"}")))`

Geez that is ugly :sweat_smile: 

That is a JSON string with escaped JSON inside. Might as well have packed everything in base64 encoded XML as well :roll_eyes:

**Comment by bonomat** [comment]
> Introducing two separate states just shifts the problem to a different state. We can make the state more fine granular here, but I don't think it would help.
> If we don't wait for confirmations on redeem at all the problem is solved.
> If we remove the timeout on waiting for confirmations the problem is solved as well.
> If we wait for a confirmations, and that waiting is bound by a timeout then the problem will exist no matter in what state - we would then have to consider re-publishing the transaction.

I think adding an additional state will help in this case: 

Once we've seen `TxRedeem` in the mempool we should invalidate the `TxCancel`/`TxPunish` path and hence the second error _should not_ appear anymore. 

Side note: once we have the other protocol this opens up a scenario where the ASB re-uses the redeem UTXO in another swap directly and allows for bumping the fees per CPFP. 

It won't solve the other two problems though which need to be handled separately: 
1. the timeout issue as you pointed out. 
2. the issue of a transaction falling out of the mempool.

**Comment by da-kami** [comment]
> > Introducing two separate states just shifts the problem to a different state. We can make the state more fine granular here, but I don't think it would help.
> > If we don't wait for confirmations on redeem at all the problem is solved.
> > If we remove the timeout on waiting for confirmations the problem is solved as well.
> > If we wait for a confirmations, and that waiting is bound by a timeout then the problem will exist no matter in what state - we would then have to consider re-publishing the transaction.
> 
> I think adding an additional state will help in this case:
> 
> Once we've seen `TxRedeem` in the mempool we should invalidate the `TxCancel`/`TxPunish` path and hence the second error _should not_ appear anymore.
> 
> Side note: once we have the other protocol this opens up a scenario where the ASB re-uses the redeem UTXO in another swap directly and allows for bumping the fees per CPFP.
> 
> It won't solve the other two problems though which need to be handled separately:
> 
>     1. the timeout issue as you pointed out.
> 
>     2. the issue of a transaction falling out of the mempool.

I did some more code analysis and have to correct myself There actually is no specific other timelock. I think the problem was actually related to my laptop being offline for a while and once it came back online the subscription ran into an error and that lead to the problem.
So, there is no timeout to remove / change. We can, however, still run into this problem if redeem does not get enough confirmations before the cancel timelock expires.

So, I think you are right - if we see redeem in mempool we should not allow transitioning to `cancel` anymore, because it would not make sense. Once the ASB sees redeem in mempool so does the CLI, and that means the Monero is redeemed. That invalidates the cancel path for the ASB.
I will see to introduce another state in a follow up of https://github.com/comit-network/xmr-btc-swap/pull/490 - I don't see this a direct blocker to merging it.

**Comment by da-kami** [comment]
I updated the `Possible Fixes` section of he ticket description to reflect some of the discussion.

**Comment by da-kami** [comment]
Part of this ticket was solved by #525 
Closing in favor of https://github.com/comit-network/xmr-btc-swap/issues/550

Note that transactions falling out of mempool is not tracked. I would not invest time into this at the current point.
